web-registry/src/lib/wallet/wallet.tsx

```diff
@@ -0,0 +1,106 @@
+import WalletConnect from '@walletconnect/client';
+
+import { useAutoConnect } from './hooks/useAutoConnect';
+import { useConnect } from './hooks/useConnect';
+import { useConnectWallet } from './hooks/useConnectWallet';
+import { useDetectKeplrMobileBrowser } from './hooks/useDetectKeplrMobileBrowser';
+import { useDisconnect } from './hooks/useDisconnect';
+import { useOnAccountChange } from './hooks/useOnAccountChange';
+import { useWalletConnectCallback } from './hooks/useWalletConnectCallback';
+import { useWalletConnectFinalize } from './hooks/useWalletConnectFinalize';
+import { emptySender } from './wallet.constants';
+import { ConnectParams } from './wallet.types';
+import { WalletConfig } from './walletsConfig/walletsConfig.types';
+
+export interface Wallet {
+  offlineSigner?: OfflineSigner;
+  address: string;
+  shortAddress: string;
+}
+
+declare global {
+  // eslint-disable-next-line @typescript-eslint/no-empty-interface
+  interface Window extends KeplrWindow {}
+}
+
+export type WalletContextType = {
+  wallet?: Wallet;
+  loaded: boolean;
+  connect?: (params: ConnectParams) => Promise<void>;
+  disconnect?: () => void;
+  connectionType?: string;
+  error?: unknown;
+  walletConnectUri?: string;
+};
+
+const WalletContext = createContext<WalletContextType>({
+  loaded: false,
+});
+
+export const WalletProvider: React.FC = ({ children }) => {
+  // Because initiating the wallet is asyncronous, when users enter the app, the wallet is seen as not loaded.
+  // This is being used so that we display the "connect wallet" or the connected wallet address
+  // only once we know what's the actual wallet connection status.
+  const [loaded, setLoaded] = useState<boolean>(false);
+  const [wallet, setWallet] = useState<Wallet>(emptySender);
+  const [connectionType, setConnectionType] = useState<string | undefined>(
+    undefined,
+  );
+  const [walletConnect, setWalletConnect] = useState<
+    WalletConnect | undefined
+  >();
+  const walletConfigRef = useRef<WalletConfig | undefined>();
+  const [error, setError] = useState<unknown>(undefined);
+  const [walletConnectUri, setWalletConnectUri] = useState<
+    string | undefined
+  >();
+
+  const onQrCloseCallback = useRef<() => void>();
+
+  const disconnect = useDisconnect({
+    setConnectionType,
+    setWallet,
+    setWalletConnect,
+    setWalletConnectUri,
+    walletConfigRef,
+    walletConnect,
+  });
+
+  const connectWallet = useConnectWallet({
+    onQrCloseCallback,
+    setWallet,
+    setWalletConnect,
+    setWalletConnectUri,
+    walletConfigRef,
+  });
+
+  const connect = useConnect({ connectWallet, setConnectionType, setError });
+
+  useAutoConnect({ connectWallet, setError, setLoaded });
+  useOnAccountChange({ connectWallet, wallet });
+  useDetectKeplrMobileBrowser({ connectWallet, loaded, wallet });
+  useWalletConnectCallback({ onQrCloseCallback, walletConnectUri });
+  useWalletConnectFinalize({ setWallet, walletConfigRef, walletConnect });
```

Comment:
one thing i'm unsure of that would be nice to understand is why we have this separate finalize action, and why we need this one, when we also have a finalize action inside of `useConnectWallet`.

Reply:
We need a separate effect for the wallet connect finalize action because it happens from a listener event (while on the desktop we can do the finalize directly).